src/main/sensors/battery.c

```diff
@@ -227,15 +236,36 @@ static void batteryUpdateVoltageState(void)
     switch (voltageState) {
         case BATTERY_OK:
             if (voltageMeter.filtered <= (batteryWarningVoltage - batteryConfig()->vbathysteresis)) {
-                voltageState = BATTERY_WARNING;
+                if (start_t == 0) {
+                    start_t = millis();
+                }
+                end_t = millis();
+                total_t = (end_t - start_t)/1000;
+                if (total_t >= batteryConfig()->batteryDurationForWarning) {
+                    voltageState = BATTERY_WARNING;
+                    start_t = 0;
+                }
+            } else {
+                start_t = 0;
             }
             break;
 
         case BATTERY_WARNING:
             if (voltageMeter.filtered <= (batteryCriticalVoltage - batteryConfig()->vbathysteresis)) {
-                voltageState = BATTERY_CRITICAL;
+                if (start_t == 0) {
+                    start_t = millis();
+                }
+                end_t = millis();
+                total_t = (end_t - start_t)/1000;
+                if (total_t >= batteryConfig()->batteryDurationForCrit) {
+                    voltageState = BATTERY_CRITICAL;
+                    start_t = 0;
+                }
             } else if (voltageMeter.filtered > batteryWarningVoltage) {
                 voltageState = BATTERY_OK;
+                start_t = 0;
```

Comment:
This will be easier to read if you make it

```
} else {
    if (voltageMeter.filtered > batteryWarningVoltage) {
￼	    voltageState = BATTERY_OK;
    }
    start_t = 0;
}
```